torchvision/models/regnet.py

```diff
@@ -0,0 +1,688 @@
+        group_widths = [self.group_width] * num_stages
+
+        # Adjust the compatibility of stage widths and group widths
+        stage_widths, group_widths = self._adjust_widths_groups_compatibilty(
+            stage_widths, bottleneck_multipliers, group_widths
+        )
+
+        return zip(
+            stage_widths, strides, stage_depths, group_widths, bottleneck_multipliers
+        )
+
+    @staticmethod
+    def _adjust_widths_groups_compatibilty(
+            stage_widths: List[int], bottleneck_ratios: List[float],
+            group_widths: List[int]) -> Tuple[List[int], List[int]]:
+        """
+        Adjusts the compatibility of widths and groups,
+        depending on the bottleneck ratio.
+        """
+        # Compute all widths for the current settings
+        widths = [int(w * b) for w, b in zip(stage_widths, bottleneck_ratios)]
+        group_widths_min = [min(g, w_bot) for g, w_bot in zip(group_widths, widths)]
+
+        # Compute the adjusted widths so that stage and group widths fit
+        ws_bot = [_make_divisible(w_bot, g) for w_bot, g in zip(widths, group_widths_min)]
+        stage_widths = [int(w_bot / b) for w_bot, b in zip(ws_bot, bottleneck_ratios)]
+        return stage_widths, group_widths_min
+
+
+def _make_blocks(
+    stem_width: int,
+    params: BlockParams,
+    norm_layer: Callable[..., nn.Module],
+    activation: Callable[..., nn.Module],
+    block_type: Callable[..., nn.Module] = ResBottleneckBlock,
+) -> Tuple[nn.Sequential, int]:
+    current_width = stem_width
+
+    blocks = []
+    for i, (
+        width_out,
+        stride,
+        depth,
+        group_width,
+        bottleneck_multiplier,
+    ) in enumerate(params.get_expanded_params()):
+        blocks.append(
+            (
+                f"block{i+1}",
+                AnyStage(
+                    current_width,
+                    width_out,
+                    stride,
+                    depth,
+                    block_type,
+                    norm_layer,
+                    activation,
+                    group_width,
+                    bottleneck_multiplier,
+                    params.se_ratio,
+                    stage_index=i + 1,
+                ),
+            )
+        )
+
+        current_width = width_out
+    return (nn.Sequential(OrderedDict(blocks)), current_width)
+
+
+class Classifier(nn.Module):
+    def __init__(self, in_channels: int, num_classes: int = 1000) -> None:
+        super().__init__()
+        self.avgpool = nn.AdaptiveAvgPool2d((1, 1))
+        self.fc = nn.Linear(in_features=in_channels, out_features=num_classes)
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = self.avgpool(x)
+        x = x.flatten(start_dim=1)
+        x = self.fc(x)
+        return x
+
+
+class RegNet(nn.Module):
+    def __init__(
+        self,
+        stem: nn.Module,
+        blocks: nn.Module,
+        classifier: nn.Module,
```

Comment:
In the most recent commit, I change the model constructor to take layer instances instead of layer constructor. In this way, user could construct their own stem or classifier outside of model. The downside is that model constructor along cannot build the model, user would also need helper functions like `_make_stem`, `_make_blocks`.
I haven't seen many patterns like this in TorchVision, so I wanna give it a try and gather feedback.

Reply:
I saw a couple places where we assume the model takes `num_classes`: [reference script](https://github.com/datumbox/vision/blob/main/references/classification/train.py#L189), [test_model](https://github.com/datumbox/vision/blob/main/test/test_models.py#L444). classifier might have to be constructed inside model then.

Reply:
I think both ways have pros and cons and when working on a new code-base it's a matter of preference. Perhaps even what you propose can be better.

Unfortunately given that the rest of the code-base uses the other idiom, diverging can cause issues. The main issue that comes to mind is when the backbone is incorporated to Object Detection and Segmentation models. Diverging means you will need to do this differently. This increases complexity and could potentially confuse users.

To avoid delaying the PR, I would recommend switching back to callable layers and using `num_classes` similar to all other models. This will allow you to eliminate some of the helper methods/classes that you introduce.. Then open a RFC where you record pros/cons of the change and investigate how we potentially do this in a BC manner. I think the models area can use a facelift, so let's just discuss/organize this there.